web/packages/teleport/src/Roles/RoleEditor/StandardEditor/validation.test.ts

```diff
@@ -0,0 +1,167 @@
+    const model = minimalRoleModel();
+    model.metadata.labels = [{ name: 'foo', value: 'bar' }];
+    model.resources = [
+      {
+        kind: 'kube_cluster',
+        labels: [{ name: 'foo', value: 'bar' }],
+        groups: [],
+        resources: [
+          {
+            id: 'dummy-id',
+            kind: { label: 'pod', value: 'pod' },
+            name: 'res-name',
+            namespace: 'dummy-namespace',
+            verbs: [],
+          },
+        ],
+      },
+      {
+        kind: 'node',
+        labels: [{ name: 'foo', value: 'bar' }],
+        logins: [{ label: 'root', value: 'root' }],
+      },
+      {
+        kind: 'app',
+        labels: [{ name: 'foo', value: 'bar' }],
+        awsRoleARNs: ['some-arn'],
+        azureIdentities: ['some-azure-id'],
+        gcpServiceAccounts: ['some-gcp-acct'],
+      },
+      {
+        kind: 'db',
+        labels: [{ name: 'foo', value: 'bar' }],
+        roles: [{ label: 'some-role', value: 'some-role' }],
+        names: [],
+        users: [],
+      },
+      {
+        kind: 'windows_desktop',
+        labels: [{ name: 'foo', value: 'bar' }],
+        logins: [],
+      },
+    ];
+    model.rules = [
+      {
+        id: 'dummy-id',
+        resources: [{ label: ResourceKind.Node, value: ResourceKind.Node }],
+        verbs: [{ label: '*', value: '*' }],
+      },
+    ];
+    const result = validateRoleEditorModel(model, undefined, undefined);
+    expect(result.metadata.valid).toBe(true);
+    expect(validity(result.resources)).toEqual([true, true, true, true, true]);
+    expect(validity(result.rules)).toEqual([true]);
+  });
+
+  test('invalid metadata', () => {
+    const model = minimalRoleModel();
+    model.metadata.name = '';
+    const result = validateRoleEditorModel(model, undefined, undefined);
+    expect(result.metadata.valid).toBe(false);
+  });
+
+  test('invalid resource', () => {
+    const model = minimalRoleModel();
+    model.resources = [
+      {
+        kind: 'node',
+        labels: [{ name: 'foo', value: '' }],
+        logins: [],
+      },
+    ];
+    const result = validateRoleEditorModel(model, undefined, undefined);
+    expect(validity(result.resources)).toEqual([false]);
+  });
+
+  test('invalid access rule', () => {
+    const model = minimalRoleModel();
+    model.rules = [
+      {
+        id: 'dummy-id',
+        resources: [],
+        verbs: [{ label: '*', value: '*' }],
+      },
+    ];
+    const result = validateRoleEditorModel(model, undefined, undefined);
+    expect(validity(result.rules)).toEqual([false]);
+  });
+
+  it('reuses previously computed section results', () => {
+    const model = minimalRoleModel();
+    const result1 = validateRoleEditorModel(model, undefined, undefined);
+    const result2 = validateRoleEditorModel(model, model, result1);
+    expect(result2.metadata).toBe(result1.metadata);
+    expect(result2.resources).toBe(result1.resources);
+    expect(result2.rules).toBe(result1.rules);
+  });
```

Comment:
Nit: we are mixing `test` and `it`, I'd use only one of them.

Reply:
The purpose if `it` and `test` existing alongside each other in Jest API is to make it easier to write code that can be read like proper English, so in this case, I'm gonna leave it as is.